### ISSUE TITLE
New linter: detect self assignment

### ISSUE THREAD

**Issue by marco-carvalho** [clj-kondo/clj-kondo#2973]
- [ ] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**Is your feature request related to a problem? Please describe.**

A `set!` that assigns a target's current value back to the same target has no effect and often remains after an incomplete refactor.

```clojure
(set! *x* *x*)
(set! (.-x obj) (.-x obj))
```

**Describe the solution you'd like**

Add a `:self-assignment` linter, enabled at warning level by default, for symbol targets and supported ClojureScript field assignment forms when the target and value are structurally the same. Different targets or values should not warn.

**Describe alternatives you've considered**

Issue #2654 and `:redundant-let-binding` cover self-referential `let` bindings, not mutation. Extending that linter would mix binding and assignment semantics.

**Additional context**

I am interested in submitting a focused PR with Clojure and ClojureScript assignment forms if this approach is accepted.

**Comment by NoahTheDuke** [comment]
How often does this happen? Do you have any examples of this in the wild?

**Comment by marco-carvalho** [comment]
I found no examples in the pinned clj-kondo or Metabase regression corpora, so I do not have evidence that it happens often in Clojure code. The motivation was that the check is deterministic and low-risk, but that alone may not justify another default-on linter. I would be comfortable making it opt-in or closing the proposal.
